File: flutter/test/support/contract_emitter.dart
// Writes factory output to contracts/out/flutter/<schema>.<variant>.json so
// `node contracts/scripts/validate.mjs --out flutter` checks it with ajv, the
// same validator every other platform's output goes through.
//
//   emitContract('events-batch', 'failure-default', makeFailureBatch());
//
// SELLWILD_CONTRACT_OUT overrides the base directory (a `flutter/` directory
// is created inside it). validate.mjs and the iOS emitter read the same
// variable; Android takes the `sellwild.contracts.outDir` system property
// instead. A relative value resolves against flutter test's cwd (the package
// root); scripts/coverage/flutter.sh makes it absolute first.

import 'dart:convert';
import 'dart:io';

import 'fixtures.dart';

final RegExp _schemaName = RegExp(r'^[a-z0-9][a-z0-9-]*$');
final RegExp _variantName = RegExp(r'^[a-z0-9][a-z0-9_-]*$');

/// <contracts>/out/flutter, or $SELLWILD_CONTRACT_OUT/flutter. Pure, so tests
/// can check the rule.
String resolveContractOutPath({
  required String contractsRoot,
  Map<String, String> environment = const {},
}) {
  final override = environment['SELLWILD_CONTRACT_OUT'];
  final base =
      override != null && override.isNotEmpty ? override : '$contractsRoot/out';
  return '$base/flutter';
}

Directory get contractOutDir => Directory(resolveContractOutPath(
      contractsRoot: contractsDir.path,
      environment: Platform.environment,
    ));

/// <out>/flutter-harness, next to [contractOutDir]: where the support
/// self-test's round-trip goes. scripts/coverage/flutter.sh validates it with
/// `--out flutter-harness`, so it proves the emit -> ajv route without ever
/// counting as factory output in `--out flutter`.
Directory get harnessContractOutDir =>
    Directory('${contractOutDir.parent.path}/flutter-harness');

/// Writes [payload] as pretty JSON and returns the file.
///
/// Throws ArgumentError when a name is malformed (a dot in the variant would
/// break the `<schema>.<variant>` split) or when no schema file named
/// [schema] exists in [schemasDir] (default contracts/schemas), so a typo
/// cannot produce a file the validator skips. jsonEncode throws on values
/// that are not JSON.
File emitContract(
  String schema,
  String variant,
  Object? payload, {
  Directory? outDir,
  Directory? schemasDir,
}) {
  if (!_schemaName.hasMatch(schema)) {
    throw ArgumentError.value(
        schema, 'schema', 'must match ${_schemaName.pattern}');
  }
  if (!_variantName.hasMatch(variant)) {
    throw ArgumentError.value(
        variant, 'variant', 'must match ${_variantName.pattern}');
  }
  final schemas = schemasDir ?? Directory(contractsPath('schemas'));
  final schemaFile = File('${schemas.path}/$schema.schema.json');
  if (!schemaFile.existsSync()) {
    throw ArgumentError.value(schema, 'schema', 'no ${schemaFile.path}');
  }
  final text = '${const JsonEncoder.withIndent('  ').convert(payload)}\n';
  final dir = outDir ?? contractOutDir;
  dir.createSync(recursive: true);
  final file = File('${dir.path}/$schema.$variant.json');
  file.writeAsStringSync(text);
  return file;
}

File: ios/Tests/SellwildSDKTests/Support/ContractEmitter.swift
import Foundation

/// Writes values produced by iOS code (factory output, wire events, request
/// bodies) as JSON for `node contracts/scripts/validate.mjs --out ios`, which
/// checks each file against `contracts/schemas/<schema>.schema.json`.
///
///     try ContractEmitter.emit(eventDict, schema: "events-batch", variant: "client-failure")
///
/// writes `contracts/out/ios/events-batch.client-failure.json`. With
/// `SELLWILD_CONTRACT_OUT` set, files go to `$SELLWILD_CONTRACT_OUT/ios/`
/// instead, matching the Android and Flutter emitters. Tests run in the
/// simulator, so `scripts/coverage/ios.sh` forwards the variable as
/// `TEST_RUNNER_SELLWILD_CONTRACT_OUT`.
///
/// Swift gets no schema library on purpose: a Package.swift dependency, even
/// a test-only one, enters partners' SPM resolution.
enum ContractEmitter {

    enum Failure: Error, CustomStringConvertible, Equatable {
        case badSchemaName(String)
        case badVariantName(String)
        case unknownSchema(String, URL)
        case notJSON(String)

        var description: String {
            switch self {
            case .badSchemaName(let name):
                return "Contract schema name \"\(name)\" must match [a-z0-9][a-z0-9-]*"
            case .badVariantName(let name):
                return "Contract variant name \"\(name)\" must match [A-Za-z0-9][A-Za-z0-9_-]*"
            case .unknownSchema(let name, let file):
                return "No schema \"\(name)\": \(file.path) does not exist, so the validator would skip this output"
            case .notJSON(let reason):
                return "Contract output is not valid JSON: \(reason)"
            }
        }
    }

    static let environmentKey = "SELLWILD_CONTRACT_OUT"
    static let platform = "ios"
    static let schemasDirectory = Fixtures.contractsDirectory.appendingPathComponent("schemas", isDirectory: true)

    /// `$SELLWILD_CONTRACT_OUT/ios` when the variable is set and not empty,
    /// else `contracts/out/ios` in this repo.
    static func outputDirectory(environment: [String: String] = ProcessInfo.processInfo.environment) -> URL {
        let base: URL
        if let override = environment[environmentKey], !override.isEmpty {
            base = URL(fileURLWithPath: override, isDirectory: true)
        } else {
            base = Fixtures.contractsDirectory.appendingPathComponent("out", isDirectory: true)
        }
        return base.appendingPathComponent(platform, isDirectory: true)
    }

    /// Writes a `JSONSerialization` value (dictionary, array, or scalar) and
    /// returns the file.
    @discardableResult
    static func emit(
        _ object: Any, schema: String, variant: String,
        directory: URL = outputDirectory(), schemas: URL = schemasDirectory
    ) throws -> URL {
        // JSONSerialization raises an Objective-C exception, not a Swift error,
        // on values it cannot write, so check first.
        guard JSONSerialization.isValidJSONObject(object) || isScalar(object) else {
            throw Failure.notJSON("\(type(of: object)) holds a value JSONSerialization cannot write")
        }
        try checkNames(schema: schema, variant: variant, schemas: schemas)
        let bytes = try JSONSerialization.data(withJSONObject: object, options: [
            .prettyPrinted, .sortedKeys, .withoutEscapingSlashes, .fragmentsAllowed,
        ])
        return try write(bytes, schema: schema, variant: variant, directory: directory)
    }

    /// Writes an `Encodable` value and returns the file.
    @discardableResult
    static func emit<T: Encodable>(
        encodable value: T, schema: String, variant: String,
        directory: URL = outputDirectory(), schemas: URL = schemasDirectory
    ) throws -> URL {
        try checkNames(schema: schema, variant: variant, schemas: schemas)
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys, .withoutEscapingSlashes]
        return try write(try encoder.encode(value), schema: schema, variant: variant, directory: directory)
    }

    /// Writes bytes that are already JSON, such as a request body captured by
    /// `StubURLProtocol`, after checking they parse. Keys come out sorted.
    @discardableResult
    static func emit(
        jsonData: Data, schema: String, variant: String,
        directory: URL = outputDirectory(), schemas: URL = schemasDirectory
    ) throws -> URL {
        let object: Any
        do {
            object = try JSONSerialization.jsonObject(with: jsonData, options: [.fragmentsAllowed])
        } catch {
            throw Failure.notJSON((error as NSError).localizedDescription)
        }
        return try emit(object, schema: schema, variant: variant, directory: directory, schemas: schemas)
    }

    // MARK: Private

    private static func isScalar(_ object: Any) -> Bool {
        object is String || object is NSNumber || object is NSNull
    }

    /// A dot in either name would break the `<schema>.<variant>.json` split,
    /// and a schema with no file would be skipped by the validator.
    private static func checkNames(schema: String, variant: String, schemas: URL) throws {
        guard isName(schema, allowUpper: false, allowUnderscore: false) else { throw Failure.badSchemaName(schema) }
        guard isName(variant, allowUpper: true, allowUnderscore: true) else { throw Failure.badVariantName(variant) }
        let schemaFile = schemas.appendingPathComponent("\(schema).schema.json")
        guard FileManager.default.fileExists(atPath: schemaFile.path) else {
            throw Failure.unknownSchema(schema, schemaFile)
        }
    }

    private static func isName(_ name: String, allowUpper: Bool, allowUnderscore: Bool) -> Bool {
        let scalars = Array(name.unicodeScalars)
        guard let first = scalars.first else { return false }
        func word(_ c: Unicode.Scalar) -> Bool {
            ("a"..."z").contains(c) || ("0"..."9").contains(c) || (allowUpper && ("A"..."Z").contains(c))
        }
        return word(first) && scalars.dropFirst().allSatisfy { c in
            word(c) || c == "-" || (allowUnderscore && c == "_")
        }
    }

    private static func write(_ bytes: Data, schema: String, variant: String, directory: URL) throws -> URL {
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        let file = directory.appendingPathComponent("\(schema).\(variant).json")
        try (bytes + Data("\n".utf8)).write(to: file, options: .atomic)
        return file
    }
}
